File: leetcode/triangle.py
from typing import List


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        self.memo = []
        self.initMemo(triangle)
        return self.minimum(triangle, 0, 0)

    memo = []

    def initMemo(self, triangle: List[List[int]]):

        for row in triangle:
            self.memo.append([None for _ in range(len(row))])

    def minimum(self, triangle: List[List[int]], row, col) -> int:

        if row >= len(triangle) or col >= len(triangle[row]):
            return 0
        if self.memo[row][col] is not None:
            return self.memo[row][col]
        self.memo[row][col] = triangle[row][col] + min(self.minimum(triangle, row + 1, col),
                                                       self.minimum(triangle, row + 1, col + 1))
        return self.memo[row][col]


if __name__ == '__main__':
    list = [[2], [3, 4], [6, 5, 7], [4, 1, 8, 3]]
    res = Solution().minimumTotal(list)
    print(res)

File: leetcode/triangle.go
package main

import (
	"fmt"
)

func mini(i, j int) int {
	if i < j {
		return i
	}
	return j
}
func minimumTotal(triangle [][]int) int {
	newrow := []int{}
	getPrevMinimum := func(row, col int) int {
		if row == 0 || row >= len(triangle) {
			return 0
		}
		if col == len(triangle[row])-1 {
			return newrow[col-1]
		}
		if col == 0 {
			return newrow[0]
		}
		return mini(newrow[col-1], newrow[col])
	}
	m := 0
	for i, row := range triangle {
		n1 := []int{}
		m = 100000
		for j, val := range row {
			v := getPrevMinimum(i, j) + val
			n1 = append(n1, v)
			m = mini(m, v)
		}
		newrow = n1
	}
	//fmt.Println(m)
	return m
}

func main() {
	triangle := [][]int{
		{2},
		{3, 4},
		{6, 5, 7},
		{4, 1, 8, 3},
	}
	res := minimumTotal(triangle)
	fmt.Println(res)
}

//    {2},
//  {3, 4},
// {6, 5, 7},
//{4, 1, 8, 3},
